src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/chemplant/GregtechMTE_ChemicalPlant.java

```diff
@@ -592,136 +585,97 @@ public void receiveClientEvent(byte aEventID, byte aValue) {
-                return false;
-            }
-        }
-
-        // Remember last recipe - an optimization for findRecipe()
-        this.mLastRecipe = tRecipe;
-
-        if (tRecipe.mSpecialValue > this.mSolidCasingTier) {
-            return false;
-        }
+            ItemStack catalystRecipe;
+            int maxParallelCatalyst;
 
-        // checks if it has enough catalyst durability
-        ArrayList<ItemStack> tCatalysts = null;
-        int tMaxParallelCatalyst = aMaxParallelRecipes;
-        if (tCatalystRecipe != null) {
-            tCatalysts = new ArrayList<ItemStack>();
-            tMaxParallelCatalyst = getCatalysts(aItemInputs, tCatalystRecipe, aMaxParallelRecipes, tCatalysts);
-        }
-
-        if (tMaxParallelCatalyst == 0) {
-            return false;
-        }
-
-        GT_ParallelHelper helper = new GT_ParallelHelper().setRecipe(tRecipe).setItemInputs(aItemInputs)
-                .setFluidInputs(aFluidInputs).setAvailableEUt(tEnergy).setMaxParallel(tMaxParallelCatalyst)
-                .enableConsumption().enableOutputCalculation().setController(this);
+            @NotNull
+            @Override
+            protected CheckRecipeResult validateRecipe(@NotNull GT_Recipe recipe) {
+                if (recipe.mSpecialValue > mSolidCasingTier) {
+                    return CheckRecipeResultRegistry.insufficientMachineTier(recipe.mSpecialValue);
+                }
+                // checks if it has a catalyst
 
-        if (batchMode) {
-            helper.enableBatchMode(128);
-        }
+                boolean needsCalayst = false;
+                for (ItemStack item : recipe.mInputs) {
+                    if (ItemUtils.isCatalyst(item)) {
+                        needsCalayst = true;
+                        break;
+                    }
+                }
+                if (needsCalayst) {
+                    catalystRecipe = findCatalyst(inputItems, recipe.mInputs);
+                    if (catalystRecipe == null || mCatalystBuses.size() != 1) {
+                        return SimpleCheckRecipeResult.ofFailure("no_catalyst");
+                    }
+                }
 
-        helper.build();
+                // checks if it has enough catalyst durability
+                maxParallelCatalyst = maxParallel;
+                if (catalystRecipe != null) {
+                    maxParallelCatalyst = getCatalysts(inputItems, catalystRecipe, maxParallel);
+                }
+                maxParallel = Math.min(maxParallel, maxParallelCatalyst);
+                return CheckRecipeResultRegistry.SUCCESSFUL;
+            }
 
-        if (helper.getCurrentParallel() == 0) {
-            return false;
-        }
+            @NotNull
+            @Override
+            protected GT_ParallelHelper createParallelHelper(@NotNull GT_Recipe recipe) {
+                return super.createParallelHelper(recipe)
+                        .enableBatchMode(isCatalystDamageable() ? maxParallelCatalyst : batchSize);
```

Comment:
Actually this won't work well because
- Argument `aBatchModifier` multiplies final parallel: `mCurrentParallel * mBatchModifier`. Currently this code allows more parallel.
- `getCatalysts` gets currently remaining durability, but `damageCatalyst` might allow more durability because of chance based damaging. Currently this code allows less parallel.

One solution came to my mind is adding `consumeInput` method to parallel helper and allowing overriding. That way we can check catalyst damaging every time input is consumed.